src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt

```diff
@@ -0,0 +1,112 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.cargo.runconfig.test
+
+import com.google.gson.*
+import com.intellij.execution.testframework.TestConsoleProperties
+import com.intellij.execution.testframework.sm.ServiceMessageBuilder
+import com.intellij.execution.testframework.sm.runner.OutputToGeneralTestEventsConverter
+import com.intellij.openapi.util.Key
+import jetbrains.buildServer.messages.serviceMessages.*
+
+data class LibtestSuiteMessage(
+    val type: String,
+    val event: String
+) {
+    companion object {
+        fun fromJson(json: JsonObject): LibtestSuiteMessage? {
+            if (json.getAsJsonPrimitive("type")?.asString != "suite") {
+                return null
+            }
+
+            return Gson().fromJson(json, LibtestSuiteMessage::class.java)
+                ?: error("Failed to parse LibtestSuiteMessage from $json")
+        }
+    }
+}
+
+data class LibtestTestMessage(
+    val type: String,
+    val event: String,
+    val name: String
+) {
+    companion object {
+        fun fromJson(json: JsonObject): LibtestTestMessage? {
+            if (json.getAsJsonPrimitive("type")?.asString != "test") {
+                return null
+            }
+
+            return Gson().fromJson(json, LibtestTestMessage::class.java)
+                ?: error("Failed to parse LibtestTestMessage from $json")
+        }
+    }
+}
+
+class CargoTestEventsConverter(consoleProperties: TestConsoleProperties, val testFrameworkName: String) :
+    OutputToGeneralTestEventsConverter(testFrameworkName, consoleProperties) {
+    private val parser = JsonParser()
+    private var first = true
+    override fun processServiceMessages(text: String, outputType: Key<*>, visitor: ServiceMessageVisitor): Boolean {
+        try {
+            if(first){
+                visitor.visitCompilationStarted(CompilationStarted("rustc"))
+                first = false
+                return true
+            }
+            if (text.trim().startsWith("finished", true)){
+                visitor.visitCompilationFinished(CompilationFinished("rustc"))
+                return true
+            }
+            val jsonElement = parser.parse(text)
+            if (!jsonElement.isJsonObject) return false
+            val suiteMessage = LibtestSuiteMessage.fromJson(jsonElement.asJsonObject)
+            if (suiteMessage == null) {
+                return handleTestMessage(jsonElement, visitor)
+            } else {
+                return handleSuiteMessage(suiteMessage, visitor)
+            }
+        } catch (es: JsonSyntaxException) {
+            return false
+        }
+    }
+
+    private fun handleSuiteMessage(suiteMessage: LibtestSuiteMessage, visitor: ServiceMessageVisitor): Boolean {
+        when (suiteMessage.event) {
+            "started" -> visitor.visitTestSuiteStarted(TestSuiteStarted(testFrameworkName))
```

Comment:
Maybe better construct message via `ServiceMessageBuilder` and call `super. processServiceMessages`?
I don't sure that it's the best way but as I can see the similar implementations for different C++ test frameworks use such approach